ostd/src/io_mem.rs

```diff
@@ -51,29 +51,20 @@ impl IoMem {
         self.limit
     }
 
-    /// Resizes the I/O memory region to the new `range`.
+    /// Slices the `IoMem`, returning another `IoMem` representing the subslice.
     ///
-    /// # Errors
+    /// # Panics
     ///
-    /// Returns an error if the new `range` is not within the current range.
-    pub fn resize(&mut self, range: Range<Paddr>) -> Result<()> {
-        let start_vaddr = paddr_to_vaddr(range.start);
-        let virtual_end = self
-            .virtual_address
-            .checked_add(self.limit)
-            .ok_or(Error::Overflow)?;
-        if start_vaddr < self.virtual_address || start_vaddr >= virtual_end {
-            return Err(Error::InvalidArgs);
-        }
-        let end_vaddr = start_vaddr
-            .checked_add(range.len())
-            .ok_or(Error::Overflow)?;
-        if end_vaddr <= self.virtual_address || end_vaddr > virtual_end {
-            return Err(Error::InvalidArgs);
+    /// This method will panic if the range is empty or out of bounds.
+    pub fn slice(&self, range: Range<usize>) -> IoMem {
```

Comment:
I agree that taking `Range<usize>` seems to be better, but I still think taking `&mut self` is better than taking `&self` and returning `Self`.

```rust
impl IoMem {
    pub fn slice(&self, range: Range<usize>) -> IoMem {
        self.clone().resize(range)
    }

    pub fn resize(&mut self, range: Range<usize>) { ... }
}
```

As you can see, the `slice` semantics can be easily implemented with `resize`, but not the other way around.

Reply:
> As you can see, the `slice` semantics can be easily implemented with `resize`, but not the other way around.

I don't think so. The following code will compile ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4bd3ff7d7403a5ad9179e2016d263578))
```rust
impl IoMem {
    pub fn slice(&self, _range: Range<usize>) -> Self {
        todo!();
    }

    pub fn resize(&mut self, range: Range<usize>) {
        *self = self.slice(range);
    }
}
```

Reply:
Fair enough. I won't insist on this point.